kexdh.c

```diff
@@ -42,65 +44,115 @@ kex_dh_keygen(struct kex *kex)
 {
 	switch (kex->kex_type) {
 	case KEX_DH_GRP1_SHA1:
-		kex->dh = dh_new_group1();
+		kex->pkey = dh_new_group1();
 		break;
 	case KEX_DH_GRP14_SHA1:
 	case KEX_DH_GRP14_SHA256:
-		kex->dh = dh_new_group14();
+		kex->pkey = dh_new_group14();
 		break;
 	case KEX_DH_GRP16_SHA512:
-		kex->dh = dh_new_group16();
+		kex->pkey = dh_new_group16();
 		break;
 	case KEX_DH_GRP18_SHA512:
-		kex->dh = dh_new_group18();
+		kex->pkey = dh_new_group18();
 		break;
 	default:
 		return SSH_ERR_INVALID_ARGUMENT;
 	}
-	if (kex->dh == NULL)
+	if (kex->pkey == NULL)
 		return SSH_ERR_ALLOC_FAIL;
-	return (dh_gen_key(kex->dh, kex->we_need * 8));
+	return (dh_gen_key(kex->pkey, kex->we_need * 8));
 }
 
 int
 kex_dh_compute_key(struct kex *kex, BIGNUM *dh_pub, struct sshbuf *out)
 {
 	BIGNUM *shared_secret = NULL;
+	const BIGNUM *p, *g;
+	EVP_PKEY *dh_pkey = NULL;
+	EVP_PKEY_CTX *ctx = NULL;
 	u_char *kbuf = NULL;
 	size_t klen = 0;
-	int kout, r;
+	int r = 0;
+	DH *dh_peer = NULL;
+	BIGNUM *copy_p = NULL, *copy_g = NULL, *copy_pub = NULL;
 
 #ifdef DEBUG_KEXDH
 	fprintf(stderr, "dh_pub= ");
 	BN_print_fp(stderr, dh_pub);
 	fprintf(stderr, "\n");
 	debug("bits %d", BN_num_bits(dh_pub));
-	DHparams_print_fp(stderr, kex->dh);
+	DHparams_print_fp(stderr, EVP_PKEY_get0_DH(kex->pkey));
 	fprintf(stderr, "\n");
 #endif
 
-	if (!dh_pub_is_valid(kex->dh, dh_pub)) {
+	if (!dh_pub_is_valid(kex->pkey, dh_pub)) {
 		r = SSH_ERR_MESSAGE_INCOMPLETE;
 		goto out;
 	}
-	klen = DH_size(kex->dh);
-	if ((kbuf = malloc(klen)) == NULL ||
+
+	if ((dh_peer = DH_new()) == NULL) {
+		error_f("allocate pkey/dh_peer failed");
+		r = SSH_ERR_ALLOC_FAIL;
+		goto out;
+	}
+
+	DH_get0_pqg(EVP_PKEY_get0_DH(kex->pkey), &p, NULL, &g);
+	if ((copy_p = BN_dup(p)) == NULL ||
+	    (copy_g = BN_dup(g)) == NULL ||
+	    DH_set0_pqg(dh_peer, copy_p, NULL, copy_g) != 1) {
+		error_f("copy pkey p/g failed");
+		r = SSH_ERR_LIBCRYPTO_ERROR;
+		goto out;
+	}
+	copy_p = copy_g = NULL;
+
+	if ((copy_pub = BN_dup(dh_pub)) == NULL ||
+	    DH_set0_key(dh_peer, copy_pub, NULL) != 1) {
+		error_f("assemble peer pubkey failed");
+		r = SSH_ERR_LIBCRYPTO_ERROR;
+		goto out;
+	}
+	copy_pub = NULL;
+
+	if ((dh_pkey = EVP_PKEY_new()) == NULL ||
+	    EVP_PKEY_set1_DH(dh_pkey, dh_peer) != 1 ||
+	    (ctx = EVP_PKEY_CTX_new(kex->pkey, NULL)) == NULL) {
```

Comment:
I would move the allocation of 'ctx' to the next if() to make it clear where this is from